libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialogBot.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Generators;
+using Microsoft.Bot.Builder.Dialogs.Declarative.Resources;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// An <see cref="IBot"/> implementation that hosts an <see cref="AdaptiveDialog"/>.
+    /// </summary>
+    public class AdaptiveDialogBot : IBot
+    {
+        private const string DefaultLocaleKey = "defaultLocale";
+        private const string RootDialogKey = "defaultRootDialog";
+
+        private const string DefaultLocale = "en-US";
+        private const string DefaultLg = "main.lg";
+
+        private static ConcurrentDictionary<ResourceExplorer, LanguageGeneratorManager> _languageGeneratorManagers = new ConcurrentDictionary<ResourceExplorer, LanguageGeneratorManager>();
+
+        private readonly string _defaultLocale;
+        private readonly string _rootDialogId;
+
+        private readonly ILogger _logger;
+        private readonly ResourceExplorer _resourceExplorer;
+        private readonly IStorage _storage;
+        private readonly BotFrameworkClient _botFrameworkClient;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveDialogBot"/> class.
+        /// </summary>
+        /// <param name="configuration">An <see cref="IConfiguration"/> instance.</param>
+        /// <param name="logger">An <see cref="ILogger"/> instance.</param>
+        /// <param name="storage">The <see cref="IStorage"/> implementation to use for this <see cref="Dialog"/>.</param>
+        /// <param name="resourceExplorer">The Bot Builder <see cref="ResourceExplorer"/> to load the <see cref="Dialog"/> from.</param>
+        /// <param name="botFrameworkClient">A <see cref="BotFrameworkClient"/> for making calls to Bot Builder Skills.</param>
+        public AdaptiveDialogBot(
+            IConfiguration configuration,
+            ILogger<AdaptiveDialogBot> logger,
+            ResourceExplorer resourceExplorer,
+            IStorage storage = null,
```

Comment:
Why not make storage required and force the dev to make an explicit decision?

Reply:
Firstly, it was a goal to try and reduce the bootstrap costs - basically everything falls back to working from memory if you don't specific storage.

Secondly, with classic coded Dialogs there is no scenario other than testing for MemoryStorage. However, with AdaptiveDialogs you could easily have a Dialog that completed immediately but still be meaningful. Using MemoryStorage for that scenario would make sense. In other words, in that case, what you really have is a JSON driven ActivityHandler.